drivers/serial/pty.c

```diff
@@ -481,133 +490,27 @@ static ssize_t pty_read(FAR struct file *filep, FAR char *buffer, size_t len)
-
-              sched_lock();
-
-              /* Check how many bytes are waiting in the pipe */
-
-              ret = file_ioctl(&dev->pd_src, FIONREAD,
-                               (unsigned long)((uintptr_t)&nsrc));
-              if (ret < 0)
-                {
-                  sched_unlock();
-                  ntotal = ret;
-                  break;
-                }
-
-              /* Break out of the loop and return ntotal if the pipe is
-               * empty.  This is another race:  There fifo was empty when we
-               * called file_ioctl() above, but it might not be empty right
-               * now.  Losing that race should not lead to any bad behaviors,
-               * however, we the caller will get those bytes on the next
-               * read.
-               */
-
-              if (nsrc < 1)
-                {
-                  sched_unlock();
-                  break;
-                }
-
-              /* Read one byte from the source the byte.  This should not
-               * block.
-               */
-
-              nread = file_read(&dev->pd_src, &ch, 1);
-              sched_unlock();
-            }
-          else
-            {
-              /* Read one byte from the source the byte.  This call will
-               * block if the source pipe is empty.
-               *
-               * REVISIT: Should not block if the oflags include O_NONBLOCK.
-               * How would we ripple the O_NONBLOCK characteristic to the
-               * contained source pipe?  file_fcntl()?  Or FIONREAD? See the
-               * TODO comment at the top of this file.
-               */
-
-              nread = file_read(&dev->pd_src, &ch, 1);
-            }
-
-          /* Check if file_read was successful */
-
-          if (nread < 0)
-            {
-              ntotal = nread;
-              break;
-            }
-
           /* Perform input processing */
 
           /* \n -> \r or \r -> \n translation? */
 
-          if (ch == '\n' && (dev->pd_iflag & INLCR) != 0)
+          if (buffer[i] == '\n' && (dev->pd_iflag & INLCR) != 0)
             {
-               ch = '\r';
+               buffer[i++] = '\r';
             }
-          else if (ch == '\r' && (dev->pd_iflag & ICRNL) != 0)
+          else if (buffer[i] == '\r' && (dev->pd_iflag & ICRNL) != 0)
             {
-              ch = '\n';
+              buffer[i++] = '\n';
             }
-
-          /* Discarding \r ?  Print character if (1) character is not \r or
-           * if (2) we were not asked to ignore \r.
-           */
-
-          if (ch != '\r' || (dev->pd_iflag & IGNCR) == 0)
+          else if (buffer[i] != '\r' && (dev->pd_iflag & IGNCR) != 0)
             {
-              /* Transfer the (possibly translated) character and update the
-               * count of bytes transferred.
-               */
-
-              *buffer++ = ch;
-              ntotal++;
+              memmove(&buffer[i], &buffer[i + 1], --ntotal - i);
```

Comment:
I truly believe that this will work, but I will still make a rework proposal:
```
bool ignore = (dev->pd_iflag & IGNCR) == 0;
FAR char *wp = buffer;
size_t i;

for (i = 0, j = 0; i < ntotal; i++)
  {
    if (buffer[i] == '\n' && (dev->pd_iflag & INLCR) != 0)
      {
        ignore = true;
      }
    else if (buffer[i] == '\r' && (dev->pd_iflag & ICRNL) != 0)
      {
        buffer[i] = '\n';
      }

    /* Discarding \r ? Print character if (1) character is not \r or
     * if (2) we were not asked to ignore \r.
     */

     if (!ignore)
       {
         *wp++ = buffer[i];
       }
  }
```

Reply:
Done.